### PR TITLE
Date provider changes

### PR DESCRIPTION
**Pull request by dephillipsmichael**
These changes add automatic activity or survey finish logic for "finishedOn" and "startedOn" variables. It also makes some minor organization changes to translateActivities to make it easier to use by sub-classes.

### CODE REVIEW

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java [hunk 1]
```diff
@@ -579,6 +590,24 @@ public void uploadTaskResult(Context context, @NonNull TaskResult taskResult) {
             }
         }
 
+        if (lastLoadedTaskGuid == null) {
+            logger.error("lastLoadedTaskGuid must be set for this task to complete and " +
+                    "set finishedOn and startedOn dates, and clientData on the bridge server");
+        } else {
+            // TODO: make GUID available in ScheduledActivity constructor and get rid of this gson nonsense
```
Comment (dephillipsmichael) [inline researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:597]: I don't have permission or really know how to change the swagger auto-generated code, but we need to be able to set a ScheduledActivity's GUID (it is a private member variable) somehow without doing this GSON reflection work-around.

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java [hunk 2]
```diff
@@ -579,6 +590,24 @@ public void uploadTaskResult(Context context, @NonNull TaskResult taskResult) {
...
+            if (taskResult.getStartDate() != null) {
+                activity.setStartedOn(new DateTime(taskResult.getStartDate()));
+            }
+            if (taskResult.getEndDate() != null) {
+                activity.setFinishedOn(new DateTime(taskResult.getEndDate()));
+            }
+            bridgeManagerProvider.getActivityManager().updateActivity(activity).subscribe(message -> {
+                logger.info("Update activity success " + message);
+            }, throwable -> logger.error(throwable.getLocalizedMessage()));
+        }
```
Comment (DwayneJengSage) [inline researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:609]: Is it worth adding unit tests for this?

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java [hunk 3]
```diff
@@ -654,6 +688,10 @@ private SchedulesAndTasksModel translateActivities(@NonNull ScheduledActivityLis
                 }
                 tsm.taskIsOptional = sa.getPersistent();
                 tsm.taskType = activity.getActivityType().toString();
+                if (sa.getFinishedOn() != null) {
+                    tsm.taskFinishedOn = sa.getFinishedOn().toDate();
+                }
+                tsm.taskGUID = sa.getGuid();
```
Comment (DwayneJengSage) [inline researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:694]: Is it worth updating BridgeDataProviderTest.testLoadTasksAndSchedules() to check for these properties?

Comment (dephillipsmichael) [reply researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:694]: It does seem like that method calls translateActivities so I'll add something in there to test this.

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java (hunk 2 above)
Comment (dephillipsmichael) [reply researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:609]: I can do that, but how would I check to see if bridgeManagerProvider.getActivityManager().updateActivity gets called in the test?

Comment (DwayneJengSage) [reply researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:609]: verify(mockActivityManager).updateActivity(eq(expectedActivity))

Comment (dephillipsmichael) [reply researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:609]: Thanks!  I added that test

File: researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java (hunk 3 above)
Comment (dephillipsmichael) [reply researchstack-sdk/src/main/java/org/sagebionetworks/bridge/researchstack/BridgeDataProvider.java:694]: Also added a test for this in testLoadTasksAndSchedules

**Pull request merged**